dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java

```diff
@@ -89,37 +87,57 @@ public void validateRelationship( ValidationErrorReporter reporter, Relationship
             return;
         }
 
-        validateRelationshipType( reporter, relationship, bundle.getPreheat() );
-        validateAutoRelationship( reporter, relationship );
-        validateBidirectionalDuplicatedRelationships( reporter, bundle, relationship );
+        boolean isValid = validateMandatoryData( reporter, relationship,
+            bundle.getPreheat().getAll( TrackerIdScheme.UID, RelationshipType.class ) );
+
+        // No need to check additional data if there are missing information on the
+        // Relationship
+        if ( isValid )
+        {
+            validateRelationshipConstraint( reporter, relationship, bundle );
+
+            validateAutoRelationship( reporter, relationship );
+
+            // validateReferences( reporter, bundle, relationship); // TODO Enrico
+        }
+
     }
 
-    private void validateBidirectionalDuplicatedRelationships( ValidationErrorReporter reporter, TrackerBundle bundle,
-        Relationship relationship )
+    private void validateRelationshipConstraint( ValidationErrorReporter reporter, Relationship relationship,
+        TrackerBundle bundle )
     {
-        List<Pair<RelationshipItem, RelationshipItem>> relationships =
-            bundle.getRelationships()
-                .stream()
-                .map( r -> new ImmutablePair<>( r.getFrom(), r.getTo() ) )
-                .collect( Collectors.toList() );
+        getRelationshipType( bundle.getPreheat().getAll( TrackerIdScheme.UID, RelationshipType.class ),
+            relationship.getRelationshipType() ).ifPresent( relationshipType -> {
+
+                validateRelationshipConstraint( "from", relationship.getFrom(), relationshipType.getFromConstraint() )
+                    .forEach( reporter::addError );
+                validateRelationshipConstraint( "to", relationship.getTo(), relationshipType.getToConstraint() )
+                    .forEach( reporter::addError );
 
-        List<ImmutablePair<RelationshipItem, RelationshipItem>> bidirectionalRelationship = Lists
-            .newArrayList( relationship )
-            .stream()
-            .filter( Relationship::isBidirectional )
-            .map( r -> new ImmutablePair<>( r.getTo(), r.getFrom() ) )
-            .collect( Collectors.toList() );
+            } );
+    }
 
-        relationships.addAll( bidirectionalRelationship );
+    private boolean validateMandatoryData( ValidationErrorReporter reporter, Relationship relationship,
+        List<RelationshipType> relationshipsTypes )
+    {
+        addErrorIfNull( relationship.getFrom(), reporter, E4007, "relationship.getFrom" );
```

Comment:
Here we use "get"+"From"/"To", while further down we just use "relationshipType", maybe we should omit the "get" part? Looks better in my opinion:

relationship.from, relationship.to, relationship.relationshiptype

Reply:
Fixed, the argument was actually ignored. The error message is shown in the PR commit message.